### ISSUE TITLE
Build failure with gobject and introspection parameters using setuptools versions 74.0.0 and upwards

### ISSUE THREAD

**Issue by n7s** [harfbuzz/harfbuzz#4972]
The meson build fails when enabling gobject and introspection if you have setuptools starting from version 74.0.0 all the way to the current version 75.6.0 where distutils.msvccompiler has been deprecated and removed. This is in Python >= 3.12 on Ubuntu Noble (24.04).

```
[107/114] Generating src/HarfBuzz-0.0.gir with a custom command (wrapped by meson to set env)
FAILED: src/HarfBuzz-0.0.gir 
env PKG_CONFIG_PATH=/home/ubuntu/harfbuzz/build/meson-uninstalled PKG_CONFIG=/usr/bin/pkg-config CC=cc /usr/bin/x86_64-linux-gnu-g-ir-scanner --quiet --no-libtool --namespace=HarfBuzz --nsversion=0.0 --warn-all --output src/HarfBuzz-0.0.gir --c-include=hb-gobject.h --cflags-begin -DHB_NO_SINGLE_HEADER_ERROR -DHAVE_GOBJECT -DHB_EXTERN= --cflags-end -I/home/ubuntu/harfbuzz/src -I/home/ubuntu/harfbuzz/build/src -I/home/ubuntu/harfbuzz/. -I/home/ubuntu/harfbuzz/build/. --filelist=/home/ubuntu/harfbuzz/build/src/libharfbuzz-gobject.so.0.61010.0.p/HarfBuzz_0.0_gir_filelist --include=GObject-2.0 --include=freetype2-2.0 --symbol-prefix=hb --symbol-prefix=hb_gobject --identifier-prefix=hb_ --pkg-export=harfbuzz-gobject --pkg-export=harfbuzz --cflags-begin -DHAVE_CONFIG_H -DHAVE_CONFIG_H -I/home/ubuntu/harfbuzz/. -I/home/ubuntu/harfbuzz/build/. -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -I/usr/include/freetype2 -I/usr/include/libpng16 -I/home/ubuntu/harfbuzz/src/. -I/home/ubuntu/harfbuzz/build/src/. -I/usr/include/gobject-introspection-1.0 --cflags-end --add-include-path=/usr/share/gir-1.0 -L/home/ubuntu/harfbuzz/build/src -lharfbuzz -L/home/ubuntu/harfbuzz/build/src --library harfbuzz-gobject -L/home/ubuntu/harfbuzz/build/src -L/home/ubuntu/harfbuzz/build/src --library harfbuzz -lglib-2.0 -lgobject-2.0 -lm -lfreetype -lgraphite2 -lgirepository-1.0
Traceback (most recent call last):
  File "/usr/bin/g-ir-scanner", line 103, in <module>
    from giscanner.scannermain import scanner_main
  File "/usr/lib/x86_64-linux-gnu/gobject-introspection/giscanner/scannermain.py", line 35, in <module>
    from giscanner.ast import Include, Namespace
  File "/usr/lib/x86_64-linux-gnu/gobject-introspection/giscanner/ast.py", line 27, in <module>
    from .sourcescanner import CTYPE_TYPEDEF, CSYMBOL_TYPE_TYPEDEF
  File "/usr/lib/x86_64-linux-gnu/gobject-introspection/giscanner/sourcescanner.py", line 25, in <module>
    from .ccompiler import CCompiler
  File "/usr/lib/x86_64-linux-gnu/gobject-introspection/giscanner/ccompiler.py", line 29, in <module>
    from distutils.msvccompiler import MSVCCompiler
ModuleNotFoundError: No module named 'distutils.msvccompiler'
[109/114] Compiling C++ object src/libharfbuzz-subset.so.0.61010.0.p/hb-subset.cc.o
ninja: build stopped: subcommand failed.
```

**Comment by devosb** [comment]
After more research and discussion with @n7s I suspect the issue would be resolved with [gobject-introspection 1.81](https://download.gnome.org/sources/gobject-introspection/1.81/gobject-introspection-1.81.2.news), which looks like it has a [fix for the change in setuptools](https://gitlab.gnome.org/GNOME/gobject-introspection/-/issues/515).

Ubuntu 24.04 had gobject-introspection 1.80 and setuptools 68.1. No problem there with the default packages, I encountered the same error since I had upgraded setuptools with pip. Ubuntu 24.10 has a [newer setuptools (74.1)](https://packages.ubuntu.com/search?keywords=python3-setuptools&searchon=names&exact=1&suite=all&section=all) which would cause the issue since the newer [gobject-introspection (1.82)](https://packages.ubuntu.com/search?keywords=gobject-introspection&searchon=names&exact=1&suite=all&section=all) does not come until Ubuntu 25.04.

I realise all of this is not a problems with the HarfBuzz library per-se, but possibly with building the library in different environments.

**Comment by n7s** [comment]
yes, we now understand it's just an issue with an external dependency and we now have a workaround. In any case, thanks a lot for all the efforts maintaining HarfBuzz, much appreciated!
